File: src/CormenIntoductionToAlgorithms.Tests/Searches.fs
module Searches

open FsCheck.Xunit
open CormenIntoductionToAlgorithms.Searches

[<Property>]
let ``Linear search`` element (array:int[]) =
    let expected = array |> Array.tryFindIndex (fun x -> x = element)
    let result = LinearSearch.Find(array, element) |> Option.ofNullable
    expected = result

[<Property>]
let ``Binary search`` element (array:int[]) =
    let contains = array |> Array.contains element
    let sortedArray = array |> Array.sort

    let result = BinarySearch.Find(sortedArray, element) |> Option.ofNullable

    match contains, result with
    | true, Some index -> sortedArray.[index] = element
    | false, None -> true
    | _ -> false

File: src/CormenIntoductionToAlgorithms.Tests/Searches.cs
using CormenIntoductionToAlgorithms.Searches;
using FsCheck.Xunit;
using CormenIntoductionToAlgorithms.Extensions;

namespace CormenIntoductionToAlgorithms.Tests;

public class Searches
{
    [Property]
    public void LinearSearchTest(int[] array, int element)
    {
        var expected = array.FindIndex(element);
        var founded = LinearSearch.Find(array, element);
        Assert.Equal(expected, founded);
    }

    // TODO
    [Property]
    public void BinarySearchTest(int[] array, int element)
    {
        Array.Sort(array);

        var expected = Array.BinarySearch(array, element);
        var founded = BinarySearch.Find(array, element);
        Assert.Equal(expected < 0 ? null : expected, founded);
    }
}
